File: cli/internal/commands/platform/meta.go
package platform

import (
	"encoding/json"
	"fmt"
	"io"
	"sort"
	"strings"

	"github.com/blackcode-switzerland/bc-issues/cli/internal/client"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/cmdutil"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/output"
	"github.com/spf13/cobra"
)

// `bk meta` mirrors GET /api/meta — the single bootstrap call an agent should
// make first. It answers "who am I, which workspaces can I write to, and which
// one is active", so an agent can pick its target BY NAME/SLUG instead of an
// opaque numeric id (the most common cause of writing to the wrong workspace).
func newMetaCmd() *cobra.Command {
	var vocabKey string
	var contractOnly bool
	cmd := &cobra.Command{
		Use:         "meta",
		Annotations: map[string]string{"routes": "GET /api/meta"},
		Short:       "Bootstrap context: who am I + every app + where each command goes",
		Long: `Print the agent bootstrap context (GET /api/meta): the authenticated
user, the active workspace, and every workspace you belong to IN THE APP THAT
ANSWERED — which is your home app, or whichever one ` + "`--app-server`" + ` names.

ONE DEPLOYMENT CANNOT ANSWER FOR ANOTHER. Since 2026-08-10 each app keeps its
own workspaces and membership in its own schema, and no app holds a grant on
another's. So:

  - ` + "`apps`" + ` is the ADDRESS BOOK — which apps exist and where they are
    deployed. It is not a grant list, and being in it does not mean you have a
    workspace there.
  - ` + "`workspaces`" + ` is populated ONLY for the app that answered. An empty
    array under another app means "NOT KNOWN HERE" — never "you have none
    there". To find out, ask that app: ` + "`bk <app> workspace list`" + `.

Pick the workspace you write to by its NAME/SLUG from the list below — not by the
numeric id (ids are opaque and easy to confuse). Then target it with
` + "`bk <app> workspace use <slug>`" + ` or a per-command ` + "`--ws <slug>`" + `. The active
workspace is only a default, not necessarily where you mean to write.

Use --ws <slug|id> to preview another workspace's context without switching.

--contract-version prints ONE line: a short fingerprint of this app's
vocabularies, limits and type lists. Poll it instead of re-reading this whole
document and the --help tree behind it — if it has not moved since your last
run, nothing in this app's contract has. It is derived from the contract itself,
so it cannot be forgotten the way a hand-bumped number can, and it does NOT
change on a redeploy that changed nothing.

--vocab <key> prints ONE vocabulary as a flat list — the values, one per line,
and a plain array under --json so it pipes. It is the answer to "what are the
valid stages?" without a parser, and it is the AUTHORITY: a flag's --help may
enumerate the values it knew when the binary was built, this reads the server.

  bk meta --vocab            list the keys this server serves
  bk meta --vocab stages     the values, with their labels
  bk meta                    unchanged

An unknown key is an error naming the keys that exist.`,
		// A positional is accepted ONLY as `--vocab`'s argument. pflag hands a
		// bare `--vocab` its NoOptDefVal and leaves the next word in args, which
		// is what makes `--vocab stages` and `--vocab=stages` both work; without
		// this check `bk meta stages` would be silently ignored, as it was.
		Args: func(cmd *cobra.Command, args []string) error {
			if len(args) == 0 {
				return nil
			}
			if !cmd.Flags().Changed("vocab") || len(args) > 1 {
				return cmdutil.Usagef("unexpected argument %q — `bk meta` takes none, and a "+
					"vocabulary key goes to `bk meta --vocab <key>`", args[0])
			}
			return nil
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, cfg, err := cmdutil.NewClientAndConfig()
			if err != nil {
				return err
			}
			meta, err := c.Meta(cmdutil.ClientWorkspaceSlug(cfg))
			if err != nil {
				return err
			}

			if contractOnly {
				return renderContractVersion(cmd, format, meta)
			}

			if cmd.Flags().Changed("vocab") {
				return renderVocab(cmd, format, meta, vocabKeyFrom(vocabKey, args))
			}

			// Re-learn the app address book (D-1). `bk meta` is the command every
			// agent runs first and the one every routing failure's hint points
			// at, so it is where the registry becomes current — and the reason a
			// stale address is always one documented command from being fixed.
			refreshRegistry(cmd, cfg, meta, c.BaseURL)

			// Machine formats print the server's payload verbatim, so every
			// dynamic block it carries (limits, media, cli, vocabulary, and
			// anything added later) reaches the agent without a CLI release.
			// Re-serialising the typed struct would silently drop them.
			var payload any = meta
			if len(meta.Raw) > 0 && (format == output.FormatJSON || format == output.FormatYAML) {
				var passthrough any
				if err := json.Unmarshal(meta.Raw, &passthrough); err == nil {
					payload = passthrough
				}
			}

			// The `routing` block is LOCAL state merged into the server's
			// payload: which server each verb tier will reach next. The server
			// cannot see it, and an agent that cannot see it either has to
			// discover a wrong-host answer by running into one.
			//
			// MERGED into the passthrough map, never replacing it, so the
			// "machine formats print the server's payload verbatim" property
			// above still holds for everything the server sent.
			if m, ok := payload.(map[string]any); ok {
				m["routing"] = buildRoutingBlock(cfg)
			}

			return output.Render(format, payload, func(w io.Writer) error {
				name := ""
				if meta.User.Name != nil {
					name = *meta.User.Name
				}
				fmt.Fprintf(w, "user:   %s <%s> (id %d, via %s)\n", name, meta.User.Email, meta.User.ID, meta.User.Via)
				if meta.ActiveWorkspace != nil {
					fmt.Fprintf(w, "active: %s (slug %s, id %d, role %s)\n",
						meta.ActiveWorkspace.Name, meta.ActiveWorkspace.Slug, meta.ActiveWorkspace.ID, meta.ActiveWorkspace.Role)
				} else {
					fmt.Fprintln(w, "active: (none — run `bk <app> workspace use <slug>`)")
				}
				fmt.Fprintln(w)

				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "\tID\tNAME\tSLUG\tROLE")
				for _, ws := range meta.Workspaces {
					mark := " "
					if ws.IsActive {
						mark = "*"
					}
					fmt.Fprintf(tw, "%s\t%d\t%s\t%s\t%s\n", mark, ws.ID, ws.Name, ws.Slug, ws.Role)
				}
				if err := tw.Flush(); err != nil {
					return err
				}

				// The apps block only appears from a Phase 4 server. Skipped rather
				// than printed empty, so an older deployment reads as "this server
				// doesn't report apps" instead of "you have no apps".
				if len(meta.Apps) > 0 {
					fmt.Fprintln(w)
					at := output.Tabwriter(w)
					fmt.Fprintln(at, "\tAPP\tNAME\tCOMMANDS\tWORKSPACES")
					for _, slug := range sortedAppSlugs(meta.Apps) {
						a := meta.Apps[slug]
						mark := " "
						if a.IsCurrent {
							mark = "*"
						}
						// "not asked" for another app, NOT "—".
						//
						// Since 2026-08-10 the server populates `workspaces` only
						// for the app answering the request: each app's membership
						// lives in its own schema and no deployment can read
						// another's. An em dash reads as "you have none there",
						// which is a different claim and one this output cannot
						// make. An agent that believes it would skip an app it can
						// actually use.
						workspaces := "not asked"
						if a.IsCurrent {
							workspaces = "—"
						}
						if len(a.Workspaces) > 0 {
							workspaces = strings.Join(a.Workspaces, ",")
						}
						fmt.Fprintf(at, "%s\t%s\t%s\t%s\t%s\n", mark, slug, a.Name, "bk "+slug+" …", workspaces)
					}
					if err := at.Flush(); err != nil {
						return err
					}
					if len(meta.Apps) > 1 {
						fmt.Fprintln(cmd.ErrOrStderr(),
							"\n\"not asked\" = this server does not know that app's workspaces, which is not the "+
								"same as you having none. Ask it: `bk <app> workspace list`.")
					}

					// Where the vocabulary lives now. The table cannot usefully
					// print an enum list, and an agent reading only this view
					// would otherwise still reach for the deprecated top-level
					// `vocabulary` key.
					//
					// Do NOT name a version here. This line used to promise the
					// top-level keys "go away in 1.12.0"; 1.12.0 shipped, the
					// keys are still served — correctly, because CLI_MIN_VERSION
					// is 1.9.1 and every binary from 1.9.1 up reads them. A
					// removal date in a string that ships inside the binary
					// cannot be corrected once it is wrong on the installed
					// copies. The server decides when they go; this only says
					// where to read instead.
					for _, slug := range sortedAppSlugs(meta.Apps) {
						if a := meta.Apps[slug]; a.IsCurrent && len(a.Vocabulary) > 0 {
							fmt.Fprintf(cmd.ErrOrStderr(),
								"\nThis app's statuses, priorities and limits: `bk meta --json` → apps.%s\n"+
									"(the top-level vocabulary/limits/media keys are deprecated — still served for older binaries)\n",
								slug)
							break
						}
					}
				}

				printRouting(w, cfg)

				if len(meta.Workspaces) == 0 {
					if len(meta.Apps) > 0 {
						// Membership without access is the quiet failure of Phase 4:
						// nothing errored, the list is just empty. Say which it is.
						fmt.Fprintln(cmd.ErrOrStderr(),
							"(no workspaces you can use this app in — run `bk <app> workspace list --all` to see "+
								"every workspace you are a member of, then ask an owner for access)")
					} else {
						fmt.Fprintln(cmd.ErrOrStderr(), "(no workspaces)")
					}
				} else {
					fmt.Fprintln(cmd.ErrOrStderr(), "\nPick your target by SLUG (e.g. `bk <app> workspace use <slug>` or `--ws <slug>`), not the id.")
				}
				return nil
			})
		},
	}
	cmd.Flags().BoolVar(&contractOnly, "contract-version", false,
		"Print ONLY this app's contract fingerprint — poll it to detect drift cheaply")
	cmd.Flags().StringVar(&vocabKey, "vocab", "",
		"Print ONE vocabulary as a flat list; pass no key to list the keys this server serves")
	// Makes a bare `bk meta --vocab` legal. The cost is that `--vocab stages`
	// leaves "stages" as a positional, which the Args validator above expects and
	// RunE reads back — see the note there.
	cmd.Flags().Lookup("vocab").NoOptDefVal = vocabAsked
	return cmd
}

// sortedAppSlugs keeps `bk meta` output stable: Go map iteration is randomised,
// and an agent diffing two runs must not see a reordered list as a change.
func sortedAppSlugs(apps map[string]client.MetaApp) []string {
	slugs := make([]string, 0, len(apps))
	for slug := range apps {
		slugs = append(slugs, slug)
	}
	sort.Strings(slugs)
	return slugs
}

// renderContractVersion prints the current app's contract fingerprint (#31).
//
// One line on stdout and nothing else, because the whole value of this command
// is that it is cheap to run and trivial to compare:
//
//	prev=$(cat .bk-contract); now=$(bk meta --contract-version)
//	[ "$prev" = "$now" ] || bk guide   # only then pay for the full re-read
//
// An EMPTY value from an older server is an error rather than an empty line.
// "" and "unchanged" must not look alike to a caller diffing two runs — that is
// the one way this command could silently tell somebody their contract is
// stable when it has no idea.
func renderContractVersion(cmd *cobra.Command, format output.Format, meta *client.Meta) error {
	app, ok := meta.Apps[meta.CurrentApp]
	if !ok || app.ContractVersion == "" {
		return fmt.Errorf(
			"this server does not serve a contract version — it predates the field; " +
				"re-read `bk meta` in full, and `bk guide` for current usage")
	}
	return output.Render(format, map[string]string{
		"app":              meta.CurrentApp,
		"contract_version": app.ContractVersion,
	}, func(w io.Writer) error {
		_, err := fmt.Fprintln(w, app.ContractVersion)
		return err
	})
}

File: packages/platform-api/src/meta.ts
// The platform half of GET /api/meta.
//
// ---------------------------------------------------------------------------
// /api/meta IS CLASS C, AND THIS IS A HELPER, NOT A FACTORY (D-20)
// ---------------------------------------------------------------------------
// Every other shared route became `factory(ctx)` or `factory(ctx, contribution)`.
// This one did not, and the difference is the point rather than an inconsistency.
//
// `/api/meta` exists to tell an agent what THIS app's vocabulary is — its
// statuses, its priorities, its limits, its entity types. That is not a
// contribution to a shared route; it is the reason the route exists. And
// docs/platform-architecture.md §7.4 is explicit that two apps' vocabularies must
// never be merged into one top-level list: an agent has to be structurally
// unable to send a sales stage to the issue tracker.
//
// So each app writes its own `/api/meta` route, and calls this for the half that
// is genuinely identical everywhere: who you are, which workspaces you can
// reach, which apps you can reach, the link vocabulary, the CLI versions.
//
// ---------------------------------------------------------------------------
// THE DEPRECATED TOP-LEVEL KEYS ARE THE CALLER'S JOB, ON PURPOSE
// ---------------------------------------------------------------------------
// `vocabulary`, `limits` and `media` are served twice — nested under
// `apps.<slug>` (current) and at the top level (deprecated 2026-08-04, removed
// after two minor releases). They MUST be the same object references, not
// copies: a divergence between the two spellings during the overlap would be
// worse than either shape alone.
//
// This helper never constructs them. It takes the app's entry as one object and
// puts it inside `apps.<slug>`; the app spreads the very same values at the top
// level. Same reference by construction, because there is only ever one object.
// If this helper built them, keeping the two in step would be a rule someone has
// to remember — which is how they drift.

import type { NextRequest } from 'next/server'
import { listAppRegistry, type User } from '@blackcode/platform-db'
import { isSuperAdmin } from '@blackcode/platform-auth'
import { CLI_LATEST_VERSION, CLI_MIN_VERSION } from '@blackcode/platform-agent'
import type { AppContext } from './app-context'
import type { WorkspaceMembershipRef } from './workspace-source'

export interface PlatformMetaOptions {
  /**
   * What the CURRENT app contributes to its own entry in `apps.<slug>` — its
   * vocabulary, limits, media rules, entity types.
   *
   * Only the current app's entry gets them, and that is not an omission: this
   * server knows its own vocabulary and has no business inventing another app's.
   * An agent reads a different app's vocabulary from that app's own /api/meta,
   * which is what `base_url` is for.
   */
  currentApp?: Record<string, unknown>
}

export interface PlatformMetaResult {
  /** Spread straight into the route's response. */
  meta: Record<string, unknown>
  /** The resolved workspace, so the app can run its own scoped queries. */
  workspace: WorkspaceMembershipRef | null
  /** The freshly-read caller. */
  user: User
}

/**
 * Assemble everything in `/api/meta` that is the same for every app.
 *
 * Returns the workspace and user alongside the payload rather than making the
 * route resolve them twice — `?ws=` resolution is exactly the logic that must
 * not be reimplemented per app, since getting it wrong leaks which workspaces
 * exist.
 */
export async function platformMetaBlock(
  app: AppContext,
  req: NextRequest,
  user: User,
  opts: PlatformMetaOptions = {}
): Promise<PlatformMetaResult> {
  // Workspace: explicit ?ws=<slug|id> override, else this app's default for the
  // caller. Both go through `app.workspaces`, which returns null if it doesn't
  // exist or the caller isn't a member (no existence leak).
  //
  // `user.active_workspace_id` is NOT read here any more, and that is a fix
  // rather than a refactor: it is one column shared by every app, so after the
  // split it names a different team depending on which app last wrote it. The
  // app's own source answers "what should I default to" — for issues that IS
  // the column, for an app with one workspace per person it is that workspace.
  const wsParam = req.nextUrl.searchParams.get('ws')
  const workspace = wsParam
    ? await app.workspaces.getForUser(wsParam, user.id)
    : await app.workspaces.getDefaultForUser(user.id)

  // Every workspace the caller belongs to in THIS app — the disambiguation list
  // an agent needs to target the right tenant by (human-readable) name/slug.
  // There is no narrower and wider list any more: a workspace this app's own
  // source reports IS one the caller can write to here.
  const [myWorkspaces, registry] = await Promise.all([
    app.workspaces.listForUser(user.id),
    listAppRegistry(app.db),
  ])

  const members = workspace ? await app.workspaces.listMembers(workspace.id) : []

  const meta: Record<string, unknown> = {
    user: {
      id: user.id,
      email: user.email,
      name: user.name,
      avatar_url: user.avatar_url,
      via: authVia(req),
      is_super_admin: isSuperAdmin(user.email),
    },
    active_workspace: workspace
      ? {
          id: workspace.id,
          name: workspace.name,
          slug: workspace.slug,
          role: workspace.member_role,
        }
      : null,
    // Pick the target by `name`/`slug` — do NOT rely on the numeric `id` to know
    // which team it is.
    workspaces: myWorkspaces.map((w) => ({
      id: w.id,
      name: w.name,
      slug: w.slug,
      role: w.member_role,
      is_active: workspace ? w.id === workspace.id : false,
    })),
    current_app: app.appSlug,
    // ── THE ADDRESS BOOK, NOT A GRANT LIST (changed 2026-08-10, Phase 5) ──────
    // Every enabled row in `platform.apps`: which apps exist and where they are
    // deployed, so the CLI can route `bk <app> …` without anyone typing a URL.
    //
    // It USED to be derived from `platform.app_access` — the apps this token
    // holds a grant for — under docs/platform-architecture.md §4.5, "an agent
    // must not discover an app its user cannot reach". Phase 5 retired that rule
    // rather than weakening it, for two measured reasons:
    //
    //   1. IT CANNOT BE DERIVED ANY MORE. Reachability now lives in each app's
    //      own membership table, and no deployment holds a Postgres grant on
    //      another app's schema (§4.3). This server can answer for itself and
    //      for nobody else.
    //   2. IT WAS ALREADY ANSWERING FALSELY. The grants named `platform.workspaces`
    //      ids for an app whose workspaces had moved to its own schema, so a
    //      brand-new issues signup was told `apps.sales.workspaces` contained
    //      their platform workspace — a workspace `apps/sales` itself 404s.
    //
    // So `workspaces` is listed ONLY for the app answering the request. An entry
    // for another app carries its address and nothing else, because its address
    // is genuinely all this server knows. Ask that app's own `/api/meta` — which
    // is what `base_url` is for, and what a CLI fan-out does.
    //
    // AN OBJECT, NOT AN ARRAY, on purpose: each app's vocabulary and limits live
    // INSIDE its entry here (§7.4). Keyed means adding one is additive; an array
    // would have to be replaced, and replacing a field agents already parse is
    // the breakage this whole sequence exists to avoid.
    apps: Object.fromEntries(
      registry.map((a) => {
        const isCurrent = a.slug === app.appSlug
        return [
          a.slug,
          {
            slug: a.slug,
            name: a.name,
            base_url: a.base_url,
            is_current: isCurrent,
            // Only ever this app's own. An empty array for another app means
            // "not known here", NOT "you have none there" — the two were
            // indistinguishable while this was grant-derived, and conflating
            // them is what produced the false claim above.
            workspaces: isCurrent ? myWorkspaces.map((w) => w.slug) : [],
            ...(isCurrent ? (opts.currentApp ?? {}) : {}),
          },
        ]
      })
    ),
    // How to address an entity in any app. A URN is built from an app's OWN
    // workspace slug and #number, so every app can print and resolve one — that
    // survived the split untouched.
    //
    // `relations` went with `bk link` on 2026-08-10: the link table is no longer
    // written by any app and `linksRoute` is unmounted everywhere, so a relation
    // vocabulary described a command that does not exist. The URN format stays
    // because putting the other end's URN in the record's own text is the
    // documented replacement for a link.
    links: {
      urn_format: 'bc:<app>:<workspace-slug>/<entity-type>/<number>',
      urn_example: `bc:${app.appSlug}:${workspace?.slug ?? '<workspace>'}/issue/1`,
    },
    // The bk versions this server advertises (also sent as X-BK-CLI-* headers).
    // Platform, not per-app: there is ONE binary and one npm package for the
    // whole platform, so an app that published its own package name here would
    // be advertising an install that does not exist.
    cli: {
      package: '@blackcode_sa/bc-issues',
      latest_version: CLI_LATEST_VERSION,
      /** Below this the CLI hard-blocks with exit 8. */
      min_version: CLI_MIN_VERSION,
      install: 'npm install -g @blackcode_sa/bc-issues',
      update: 'npm install -g @blackcode_sa/bc-issues@latest',
    },
    // Pointers only — the behaviour itself lives in `bk guide`, which ships
    // inside the binary and therefore always describes the binary in your hand.
    conventions: {
      interface:
        'This product is operated through the bk CLI. Run `bk guide` for the complete, current usage guide for your installed binary; `bk <group> <command> --help` for flags.',
      workspace_selection:
        "Before creating anything, confirm which workspace you are writing to. The `workspaces` array above lists every workspace you belong to; match the user's intent by `name`/`slug`, never by the numeric `id` (ids are opaque and easy to confuse). `active_workspace` is only a default — it is NOT necessarily where the user means to write. Set it with `bk workspace use <slug>`, or target one command with `bk --ws <slug> …`.",
      staying_current:
        'If a command that used to work now fails, run `bk skill sync` (updates your agent skill, and tells you when the binary itself is behind), then `bk changelog` for the dated record.',
    },
    members,
  }

  return { meta, workspace, user }
}

/** Which credential the caller used. Same test the app's own resolver makes. */
function authVia(req: NextRequest): 'session' | 'token' {
  const header = req.headers.get('authorization')
  return header && /^Bearer\s+/i.test(header) ? 'token' : 'session'
}
